Compositor: Resize `PinchZoom::unscaled_viewport` when viewport size changes

kongbai1996 opened the pull request:
Implement resize method for PinchZoom and update webview_renderer to call it on window size changes
Testing: No new WPT test cases passed.
Fixes: N/A

mrobinson reviewed (changes requested):
I don't think this is correct. Scroll can cause a pan of the viewport.

yezhizhen commented:
> You can see why this is problem if you run Servo on desktop, pinch zoom, and then scroll with the mouse wheel. Without the pan here, you cannot scroll to the edge of the page.

Make sense.

kongbai1996 committed:
Implement resize method for PinchZoom and update webview_renderer to call it on window size changes

- Added a `resize` method to the `PinchZoom` struct to update the unscaled viewport size based on the new webview rectangle.
- Modified the `WebViewRenderer` to invoke the `resize` method of `PinchZoom` whenever the window size changes, ensuring the pinch zoom functionality adapts to the new dimensions.

Signed-off-by: kongbai1996 <1782765876@qq.com>

yezhizhen commented on components/compositing/pinch_zoom.rs:
```diff
@@ -34,6 +34,10 @@ impl PinchZoom {
         Scale::new(self.zoom_factor)
     }
 
+    pub(crate) fn resize(&mut self, webview_rect: DeviceRect) {
```
Ok I think I understand what you are doing now. But it really makes no sense to be named as `resize` for `PinchZoom` when only changing `unscaled_viewport_size`.

Maybe something like this
```suggestion
    pub(crate) fn resize_unscaled_viewport(&mut self, webview_rect: DeviceRect) {
```

yezhizhen reviewed (approved):
Make sense!

kongbai1996 committed:
fix suggestion

Signed-off-by: kongbai1996 <1782765876@qq.com>

mrobinson reviewed (approved):
I think there's a chance this isn't going to work how you want it to if there is already some `PinchZoom` applied to the pinch zoom viewport. If possible for your usecase please just clear the entire pinch zoom (reset transform to identity and reset the zoom_factor to 1) when resizing the viewport.

yezhizhen commented:
> I think there's a chance this isn't going to work how you want it to if there is already some `PinchZoom` applied to the pinch zoom viewport. If possible for your usecase please just clear the entire pinch zoom (reset transform to identity and reset the zoom_factor to 1) when resizing the viewport.

To play safe yeah. @kongbai1996 tested for a while without resetting, seems ok.
